hazelcast/src/main/java/com/hazelcast/config/AbstractFactoryWithPropertiesConfig.java

```diff
@@ -0,0 +1,147 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.config;
+
+import java.util.Objects;
+import java.util.Properties;
+
+/**
+ * Configuration base for config types with a factory class and its properties.
+ *
+ * @param <T> final child type
+ */
+public abstract class AbstractFactoryWithPropertiesConfig<T extends AbstractFactoryWithPropertiesConfig<T>> {
+
+    private boolean enabled;
+    private String factoryClassName;
+    private Properties properties = new Properties();
+
+    /**
+     * Returns the factory class name.
+     */
+    public String getFactoryClassName() {
+        return factoryClassName;
+    }
+
+    /**
+     * Sets the factory class name.
+     */
+    public T setFactoryClassName(String factoryClassName) {
+        this.factoryClassName = factoryClassName;
+        return self();
+    }
+
+    /**
+     * Returns if this configuration is enabled.
+     *
+     * @return {@code true} if enabled, {@code false} otherwise
+     */
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    /**
+     * Enables and disables this configuration.
+     *
+     * @param enabled {@code true} to enable, {@code false} to disable
+     */
+    public T setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        return self();
+    }
+
+    /**
+     * Sets a single property.
+     *
+     * @param name  the name of the property to set
+     * @param value the value of the property to set
+     * @return the updated SSLConfig
+     * @throws NullPointerException if name or value is {@code null}
+     */
+    public T setProperty(String name, String value) {
+        properties.put(name, value);
+        return self();
+    }
+
+    /**
+     * Gets a property.
+     *
+     * @param name the name of the property to get
+     * @return the value of the property, null if not found
+     * @throws NullPointerException if name is {@code null}
+     */
+    public String getProperty(String name) {
+        return properties.getProperty(name);
+    }
+
+    /**
+     * Gets all properties.
+     *
+     * @return the properties
+     */
+    public Properties getProperties() {
+        return properties;
+    }
+
+    /**
+     * Sets the properties.
+     *
+     * @param properties the properties to set
+     * @return the updated SSLConfig
+     * @throws IllegalArgumentException if properties is {@code null}
+     */
+    public T setProperties(Properties properties) {
+        if (properties == null) {
```

Comment:
Can you use `com.hazelcast.internal.util.Preconditions#checkNotNull(T, java.lang.String)` instead? It's commonly used in some other config files. It changes the thrown exception type though to `NullPointerException`.

Reply:
I guess I can't as the `checkNotNull` throws a NPE and `setProperties` throws the IAE.

Reply:
If this was the case already in 4.0 then yes, we can't change it. But if it's newly introduced config, I'd prefer `checkNotNull` as it's commonly used in public API.

Reply:
It's the old thingy from the `SSLConfig` - the `AbstractFactoryWithPropertiesConfig` is a new shared parent class to which some methods were moved from the `SSLConfig`.